src/Host/Client/Impl/Session/RSession.cs

```diff
@@ -262,34 +266,31 @@ internal sealed class RSession : IRSession, IRCallbacks {
 
             await _initializationTcs.Task;
             initializationCts.Dispose();
+            _stopHostLock.EnqueueReset();
             lockToken.Set();
         }
 
-        public IRSessionSwitchBrokerTransaction StartSwitchingBroker() {
-            _disposeToken.ThrowIfDisposed();
-            return new BrokerTransaction(this);
-        }
+        public IRSessionSwitchBrokerTransaction StartSwitchingBroker() => !_disposeToken.IsDisposed ? new BrokerTransaction(this) : null;
 
-        public IRSessionReconnectTransaction StartReconnecting() {
-            _disposeToken.ThrowIfDisposed();
-            return new BrokerTransaction(this);
-        }
+        public IRSessionReconnectTransaction StartReconnecting() => !_disposeToken.IsDisposed ? new BrokerTransaction(this) : null;
 
         public async Task StopHostAsync() {
             _disposeToken.ThrowIfDisposed();
             await TaskUtilities.SwitchToBackgroundThread();
 
-            var lockToken = await _initializationLock.WaitAsync();
-
-            // Host wasn't started yet or host is already stopped - nothing to stop, just pass acquired lock to the next awaiter
-            if (!lockToken.IsSet) {
-                lockToken.Reset();
+            var stopToken = await _stopHostLock.WaitAsync();
+            if (stopToken.IsSet) {
                 return;
             }
 
-            ResetInitializationTcs();
+            try {
+                ResetInitializationTcs();
+                await StopHostAsync(BrokerClient, _startupInfo.Name, _host, _hostRunTask);
 
-            await StopHostAsync(BrokerClient, _startupInfo.Name, _host, _hostRunTask);
+                stopToken.Set();
+            } finally {
+                stopToken.Reset();
+            }
         }
 
         private static async Task StopHostAsync(IBrokerClient brokerClient, string hostName, RHost host, Task hostRunTask) {
```

Comment:
This one can be called from 2 places without lock (`StopHostAsync` locks but `AfterHostStarted` - not sure)  - do we care?